lib/sinon.js

```diff
@@ -7,7 +7,21 @@
  * Copyright (c) 2010-2013 Christian Johansen
  */
 "use strict";
-module.exports = require("./sinon/util/core");
+
+var match = require("./sinon/match");
+
+module.exports = exports = require("./sinon/util/core");
+
+exports.extend = require("./sinon/extend");
+exports.match = match;
+exports.typeOf = require("./sinon/typeOf");
+
+/*
+ * allow deepEqual to check equality of matchers through
+ * dependency injectection. Otherwise we get a circular
+ * dependency
+ */
+exports.deepEqual.match = match;
```

Comment:
Such a nasty business, but in order not to break backwards compatibility I would go in the direction of your upgrade function mixed with the last version. What about ending up with something like this:

```
let matcher = require('sinon/match');
let deepEqualWithoutMatcher = require('.../core/deep-equal');
let deepEqualWithMatcher = require('.../core/deep-equal').withMatcher(matcher)
```

You could keep your old and get a new version as well.